plumbing/protocol/packp/ulreq_encode.go

```diff
@@ -78,23 +68,38 @@ func (e *ulReqEncoder) encodeFirstWant() stateFn {
 }
 
 func (e *ulReqEncoder) encodeAditionalWants() stateFn {
-	for _, w := range e.sortedWants[1:] {
+	last := e.data.Wants[0]
+	for _, w := range e.data.Wants[1:] {
+		if bytes.Compare(last[:], w[:]) == 0 {
+			continue
+		}
+
 		if err := e.pe.Encodef("want %s\n", w); err != nil {
 			e.err = fmt.Errorf("encoding want %q: %s", w, err)
 			return nil
 		}
+
+		last = w
 	}
 
 	return e.encodeShallows
 }
 
 func (e *ulReqEncoder) encodeShallows() stateFn {
-	sorted := sortHashes(e.data.Shallows)
-	for _, s := range sorted {
+	plumbing.HashesSort(e.data.Shallows)
```

Comment:
This is not the correct way to solve this.

If we are going to create a new `plubing.HashesSort([]Hash)`, why not create a `plumbing.UniqueHashes([]Hash) []Hash` in addition, which is really what we want to do?

That would avoid duplicating code, improve code legibility and communicate intention better.